src/click/core.py

```diff
@@ -1022,14 +1033,21 @@ def show_help(ctx: Context, param: Parameter, value: str) -> None:
                 echo(ctx.get_help(), color=ctx.color)
                 ctx.exit()
 
-        return Option(
-            help_options,
-            is_flag=True,
-            is_eager=True,
-            expose_value=False,
-            callback=show_help,
-            help=_("Show this message and exit."),
-        )
+        # Cache the help option object in private _help_option attribute to
+        # avoid creating it multiple times. Not doing this will break the
+        # callback odering by iter_params_for_processing(), which relies on
+        # object comparison.
+        if not getattr(self, "_help_option", None):
+            self._help_option = Option(
+                help_options,
+                is_flag=True,
+                is_eager=True,
+                expose_value=False,
+                callback=show_help,
+                help=_("Show this message and exit."),
+            )
+
+        return self._help_option
```

Comment:
Can you instantiate `self._help_option: HelpOption | None = None` in the constructor so we can check for `None` here and have this typed?

Any reason we couldn't use the `HelpOption` type you introduced as well?

Reply:
Good idea on typing `_help_option`.

As for not reusing `HelpOption`, it's because I was targeting the `main` branch instead of the `stable` one. I just fixed that.